### DISCUSSION TITLE
Frequently reading single instruction from large byte array at offset in C#

### DISCUSSION THREAD

**Discussion by winscripter** [icedland/iced#649] [Q&A]
Hello,

Does Iced support an efficient way to read a single instruction at offset from a potentially large byte array (which can be as massive as gigabytes)?

My first thought was the following, done by reading a small chunk of 16 bytes starting at offset to ensure that Iced has enough data to decode any instruction, but I realized that
1. Creating a new Decoder instance and a small byte array every time creates overhead for the .NET garbage collector and potentially excessive memory usage
2. Instructions in x86/x64 don't have a fixed length unlike ARM64 or RISC-V, making the following inefficient.
```cs
using Iced.Intel;

internal sealed class Cpu(int memorySize)
{
    private readonly byte[] _memory = new byte[memorySize];

    public Instruction Decode(int offset, int bitness = 64)
    {
        byte[] slice = new byte[16];
        for (int i = 0; i < 16; i++)
        {
            slice[i] = _memory[offset + i];
        }
        var decoder = Decoder.Create(bitness, slice);
        decoder.Decode(out Instruction instruction);
        return instruction;
    }
}
```

**Comment by wtfsck** [comment]
You're using C# so you could implement your own `CodeReader` to implement your own read-a-byte logic. Pass your `CodeReader` to the `Decoder` constructor. I think the README has an example.
